### PR TITLE
Keep undo history even when file changes outside Code.

### PR DESCRIPTION
**Pull request by stringham**
Implements #2908

### CODE REVIEW

**Comment by stringham** [comment]
@alexandrudima are there plans to look at this?

**Comment by BridgeAR** [comment]
It would be great to get this in. There are quite some people waiting on this feature.

**Comment by alexandrudima** [comment]
Sorry for not replying, I've looked at this multiple times, but I still need to figure out what's the best way to approach the problem.

The fundamental flaw is the implementation of a `O(N^2)` algorithm to find a LCS between the old model lines and the new model lines. This has two problems:
* the algorithm is `O(N^2)`, where `N` is the count of lines. In real-life, `N` can be in the millions range, which is pretty bad.
* we already have an implementation for a superior runtime diffing algorithm in `vs/base/common/diff/diff.ts` -- "An O(ND) Difference Algorithm and its letiations" by Eugene W. Myers

I'm not comfortable to run either of the algorithms on the main thread, as this could take ... minutes (during which time the main thread is blocked). I'm also not comfortable to delegate the computation of the diff to a web worker thread, as that will mean twice the memory consumption and ... it will still potentially take minutes to show the changed file on disk.

I'll think today of a simpler heuristic to capture the log file scenario: e.g. a 200MB file with millions of lines to which a line is appended every N seconds. I'd like for us to at least cover that case.... The root problem here is memory consumption, if we keep 200MB in the undo stack for every file change on disk we're gonna run out of memory pretty quickly.

**TL;DR**: I'll reserve time today to figure out what's best to be done here.

**Comment by alexandrudima** [comment]
I don't see a way to get around running the diff algorithm on the main thread, so I've simply replaced the usage of the custom LCS implementation with our existing algorithm.

**Comment by alexandrudima** [comment]
Thank you @stringham for your contribution :heart:.

If all goes well with testing, this will make it into our August release!

File: src/vs/workbench/services/textfile/common/textFileEditorModel.ts [hunk 1]
```diff
@@ -413,7 +413,6 @@ export class TextFileEditorModel extends BaseTextEditorModel implements ITextFil
 		this.blockModelContentChange = true;
 		try {
 			this.updateTextEditorModel(value);
-			this.setDirty(false);
```
Comment (stringham) [inline src/vs/workbench/services/textfile/common/textFileEditorModel.ts]: @alexandrudima I'm pretty sure this change was necessary. We need the model to know that this is now the current saved state.

If after reloading the file on disk, if you press undo it shows that the file is still saved, after a redo it shows the file is unsaved, and if you attempt to close the file it prompts you to save it, even though it is up to date.

Here's a gif where I modify the file with vim, press undo, then redo: 
dirty

Comment (bpasero) [reply src/vs/workbench/services/textfile/common/textFileEditorModel.ts]: @stringham thanks we are looking into restoring that, you are right, it is needed.

Comment (alexandrudima) [reply src/vs/workbench/services/textfile/common/textFileEditorModel.ts]: @stringham :+1: thank you :heart: !
